Add Saflok Emulation Support

aaronjamt opened the pull request:
# Superseded by #474 

-----

## What's new

- Adds the ability to create and emulate Saflok hotel keycards using Mifare Classic 1k

-----
## For the reviewer

- [ ] I've uploaded the firmware with this patch to a device and verified its functionality
- [ ] I've confirmed the bug to be fixed / feature to be stable

aaronjamt committed:
Add Saflok Emulation Support

aaronjamt commented:
Marking as draft, awaiting merge of Xero firmware updates.

aaronjamt committed:
Switch to NumberInput & Cleanup

aaronjamt commented on applications/main/nfc/helpers/saflok.c:
```diff
@@ -0,0 +1,257 @@
+    }
+    sum = 255 - (sum & 0xFF);
+    return sum & 0xFF;
+}
```
A lot of this is duplicated with the Saflok parser ([applications/main/nfc/plugins/supported_cards/saflok.c](https://github.com/Next-Flip/Momentum-Firmware/blob/dev/applications/main/nfc/plugins/supported_cards/saflok.c)), can it be shared in a header file somewhere instead?

WillyJL replied:
You could move it into the helpers folder, then add it to the app api table so it's usable inside the saflok plugin, some things like mosgortrans do that too

aaronjamt replied:
I believe I did this right, let me know if there's any issues with it!

WillyJL replied:
LGTM from a quick look, well done!

aaronjamt commented on applications/main/nfc/helpers/saflok.c:
```diff
@@ -0,0 +1,257 @@
+
+        mf_classic_set_block_read(mfc_data, block, &mfc_data->block[block]);
+
+        // This is the default log header for cards with no log data
+        // 00 00 00 00 00 00 00 00   00 00 00 C1 00 00 00 00
+        if(block == 4) {
+            mfc_data->block[block].data[11] = 0xC1;
+        }
```
This works on my MT lock, but I'd like more testing to verify it works on others as well.

aaronjamt commented on applications/main/nfc/helpers/saflok.c:
```diff
@@ -0,0 +1,257 @@
+
+    // Saflok data is stored in block 1 and the first byte of block 2
+    memcpy(mfc_data->block[1].data, data, 16);
+    mfc_data->block[2].data[0] = data[16];
```
Should this be converted to a 1-byte `memcpy` to make it a bit more readable?

WillyJL replied:
Nah seems fine, 1 byte memcpy is a waste lol

aaronjamt committed:
Tiny bit more cleanup
No idea when or why I added this lol

aaronjamt committed:
Create Saflok API
This prevents code reuse of the Saflok functions that are common across both the parser and generator.

aaronjamt committed:
Improve UI/UX
Based on suggestions by @xtruan

aaronjamt committed:
Add Saflok MFULC parser
Written by @xtruan. Thanks for the contribution!

Co-authored-by: xtruan <xtruan@users.noreply.github.com>

aaronjamt commented:
I should also note that this PR should be merged first: https://github.com/flipperdevices/flipperzero-firmware/pull/4261
Edit to add: v1 of that PR was already merged, but there's a [new commit](https://github.com/flipperdevices/flipperzero-firmware/pull/4261/commits/65cd636f500d6a7205a7e9d6ccc109844dc30825) from a couple days ago that adds another function to the API.
If you don't want to merge the extra commit yet, I can revert it and make a new PR later for it.

aaronjamt commented:
Superseded by #474
